pallets/free-calls/src/lib.rs

```diff
@@ -0,0 +1,430 @@
+            + GetDispatchInfo
+            + From<frame_system::Call<Self>>
+            + IsSubType<Call<Self>>
+            + IsType<<Self as frame_system::Config>::Call>;
+
+        /// The configurations that will be used to limit the usage of the allocated quota to these
+        /// different configs.
+        #[pallet::constant]
+        type WindowsConfig: Get<Vec<WindowConfig<Self::BlockNumber>>>;
+
+        /// Filter on which calls are permitted to be free.
+        type CallFilter: Contains<<Self as Config>::Call>;
+
+        /// Weight information for extrinsics in this pallet.
+        type WeightInfo: WeightInfo;
+
+        /// A calculation strategy to convert locked tokens info to a quota.
+        type QuotaCalculationStrategy: QuotaCalculationStrategy<Self>;
+    }
+
+    /// Keeps track of each windows usage for each consumer.
+    #[pallet::storage]
+    #[pallet::getter(fn window_stats_by_consumer)]
+    pub(super) type WindowStatsByConsumer<T: Config> = StorageDoubleMap<
+        _,
+        Blake2_128Concat,
+        T::AccountId,
+        Twox64Concat,
+        // Index of the window in the list of window configurations.
+        WindowType,
+        ConsumerStats<T::BlockNumber>,
+    >;
+
+    #[pallet::event]
+    #[pallet::generate_deposit(pub (super) fn deposit_event)]
+    pub enum Event<T: Config> {
+        /// Free call was executed. [who, result]
+        FreeCallResult(T::AccountId, DispatchResult),
+    }
+
+    #[pallet::call]
+    impl<T: Config> Pallet<T> {
+
+        /// Try to execute a call using the free allocated quota. This call may not execute because
+        /// one of the following reasons:
+        ///  * Caller has no free quota set.
+        ///  * The caller has used all the allowed quota for at least one window config.
+        #[pallet::weight({
+            let boxed_call_info = call.get_dispatch_info();
+            let boxed_call_weight = boxed_call_info.weight;
+            let self_weight = <T as Config>::WeightInfo::try_free_call();
+
+            let total_weight = self_weight.saturating_add(boxed_call_weight);
+            (
+                total_weight,
+                boxed_call_info.class,
+                Pays::No,
+            )
+        })]
+        pub fn try_free_call(
+            origin: OriginFor<T>,
+            call: Box<<T as Config>::Call>,
+        ) -> DispatchResultWithPostInfo {
+            let sender = ensure_signed(origin.clone())?;
+
+            let mut actual_weight = <T as Config>::WeightInfo::try_free_call();
+
+            if T::CallFilter::contains(&call) &&
+                Self::can_make_free_call(&sender, ShouldUpdateConsumerStats::YES) {
+                // Add the current weight for the boxed call
+                actual_weight = actual_weight.saturating_add(call.get_dispatch_info().weight);
+
+                // Dispatch the call
+                let result = call.dispatch(origin);
```

Comment:
this doesn't make sense to me. You are just doing the same math here as you did above. you will get the same results.

what you want to do is actually get the result from `let result = call.dispatch(origin);` and from that get the real weight, and use that for the math

Reply:
See here: https://github.com/paritytech/substrate/blob/master/frame/utility/src/lib.rs#L218

Reply:
Right, that makes more sense. Thank you.